colibri/src/icons.rs

```diff
@@ -119,6 +138,86 @@ async fn query_token_icon_and_extension(
     None
 }
 
+async fn query_uniswap_v3_position_icon(
+    chain_id: u64,
+    token_id: &str,
+    contract_address: &str,
+    globaldb: &globaldb::GlobalDB,
+) -> Option<(Bytes, &'static str)> {
+    let blockchain = SupportedBlockchain::from_chain_id(chain_id)?;
+    let rpc_endpoints = globaldb.get_rpc_nodes(blockchain).await.ok()?;
+    let contract_address = match Address::parse_checksummed(contract_address, None) {
+        Ok(address) => address,
+        Err(_e) => {
+            return None;
+        }
+    };
+    let token_id: U256 = token_id.parse().unwrap();
+
+    // Try all RPC endpoints until one succeeds
+    for rpc_url in rpc_endpoints {
+        let provider = match rpc_url.parse().map(|url| ProviderBuilder::new().on_http(url)) {
+            Ok(provider) => provider,
+            Err(e) => {
+                error!("Failed to setup provider for {}: {}", rpc_url, e);
+                continue;
+            }
+        };
+
+        let contract = IERC721Metadata::new(contract_address, provider);
+        let token_uri = match contract.tokenURI(token_id).call().await {
+            Ok(result) => result._0,
+            Err(e) => {
+                error!("RPC call failed on {}: {}", rpc_url, e);
+                continue;
+            }
+        };
+        let base64_str = match token_uri.strip_prefix("data:application/json;base64,") {
+            Some(data) => data,
+            None => {
+                error!("Invalid token URI format from {}", rpc_url);
+                continue;
+            }
+        };
+        let json_bytes = match BASE64_STANDARD.decode(base64_str) {
+            Ok(bytes) => bytes,
+            Err(e) => {
+                error!("Failed to decode base64 JSON from {}: {}", rpc_url, e);
+                continue;
+            }
+        };
+        let json_data: serde_json::Value = match serde_json::from_slice(&json_bytes) {
+            Ok(json) => json,
+            Err(e) => {
+                error!("Failed to parse JSON from {}: {}", rpc_url, e);
+                continue;
+            }
+        };
+        let image_uri = match json_data.get("image").and_then(|v| v.as_str()) {
+            Some(uri) => uri,
+            None => {
+                error!("No 'image' field in JSON from {}", rpc_url);
+                continue;
+            }
+        };
+        let image_base64 = match image_uri.strip_prefix("data:image/svg+xml;base64,") {
+            Some(data) => data,
+            None => {
+                error!("Invalid image URI format from {}", rpc_url);
+                continue;
+            }
+        };
+        match BASE64_STANDARD.decode(image_base64) {
+            Ok(image_data) => return Some((Bytes::from(image_data), "svg")),
+            Err(e) => {
+                error!("Failed to decode uniswap v3 base64 image from {}: {}", rpc_url, e);
+                continue;
+            }
+        }
+    }
```

Comment:
this whole function is very confusing. I believe it could be written in a more clear way and with comments. Also I believe that you continue for errors that are fatal